velox/functions/lib/Re2Functions.cpp

```diff
@@ -1170,7 +1176,16 @@ PatternMetadata determinePatternKind(
     } else {
       // Record the first fixed pattern start.
       if (fixedPatternStart == -1) {
-        fixedPatternStart = iterator.currentIndex();
+        if (iterator.isEscaped()) {
```

Comment:
Nice catch. Thank you for the fix.

currentIndex() method and this logic beaks the abstraction, the promise that iterator handles escape characters transparently.

Can we maybe replace currentIndex() with currentStart() and nextStart() methods that return the start position of the current logical character and start position of the next logical character. Or something similar that doesn't require the caller to handle escape character explicitly.

Reply:
@zhztheplayer thanks for the catch!

Um, just like @mbasmanova pointed out, the key here is `currentIndex()` breaks the abstraction, current simple way of parsing the pattern is rigid so it depends on currentIndex(). I have an incoming PR which will support optimizations for more relaxed pattern like `_a_b%` which will require more flexible parsing of the pattern. In that PR I remove the dependency on `currentIndex()`, just rely on `current()` to tell me what current char is, so after collecting all the chars from `current()`, it is the `unescaped` string we are looking for.

So my suggestion is:

- I submit a PR to disable the optimization for escaped scenario for the moment.
- In my PR to support relaxed pattern, fix this issue thoroughly.

Reply:
Thanks for the suggestion.

I looked at the code but found that we don't appear to require for `nextStart()`, `currentStart()` could be enough. Let's only add one for now to keep the code simple?

Reply:
Sounds good to me. Thanks.

Reply:
> @zhztheplayer thanks for the catch!
> 
> Um, just like @mbasmanova pointed out, the key here is `currentIndex()` breaks the abstraction, current simple way of parsing the pattern is rigid so it depends on currentIndex(). I have an incoming PR which will support optimizations for more relaxed pattern like `_a_b%` which will require more flexible parsing of the pattern. In that PR I remove the dependency on `currentIndex()`, just rely on `current()` to tell me what current char is, so after collecting all the chars from `current()`, it is the `unescaped` string we are looking for.
> 
> So my suggestion is:
> 
> * I submit a PR to disable the optimization for escaped scenario for the moment.
> * In my PR to support relaxed pattern, fix this issue thoroughly.

@xumingming Hi mingming, sorry for missing your message. So would you mean to fix the issue in your incoming PR instead? If yes, it should be OK to me but the regression blocks our daily testing which is a little bit hurry for us to have a fix landed. Is it possible to use this PR then rebase your patch onto it, if it works for you?

Reply:
@zhztheplayer Sure, go ahead.